sot/eng_archive
# Improve remote access and fix sync to work on FOT Windows

This does two somewhat independent things:

- Improve remote access control by allowing for a `SKA_ACCESS_REMOTELY` user environment variable to explicitly control whether the remote server is used.  Along the way the logic and implementation were refined as noted in the improved user documentation.
- Adds unit tests of remote access (to the extent possible) and describes manually-run tests.
- Fix some path issues in the sync mechanism that only showed up on a FOT Windows laptop.

Testing using an installed version of the package.  "Limited" unit tests done with:
```
>>> import Ska.engarchive
>>> Ska.engarchive.test('-k test_sync', '-s', '-v')
>>> Ska.engarchive.test('-k test_remote_access', '-v')
```

Testing status

- [x] TA runs limited unit tests on Windows VM
- [x] TA runs manual tests on Windows VM
- [x] TA runs limited unit tests on Mac
- [x] TA runs manual test on Mac
- [x] TA runs all unit tests on linux
- [x] MB runs limited unit tests on FOT Windows laptop
- [x] MB runs manual tests on FOT Windows laptop
- [x] MB tests in https://github.com/sot/eng_archive/pull/170#issuecomment-524032307

Manual test:
- Nominal remote access on Windows:
  No env vars set (SKA, ENG_ARCHIVE, SKA_ACCESS_REMOTELY).
  Confirm that remote access is enabled and works by fetching an MSID::
```
    import os
    os.environ.pop('SKA', None)
    os.environ.pop('ENG_ARCHIVE', None)
    os.environ.pop('SKA_ACCESS_REMOTELY', None)
    from Ska.engarchive import fetch, remote_access
    fetch.add_logging_handler()
    assert remote_access.access_remotely is True
    dat = fetch.Msid('tephin', '2018:001', '2018:010')
    print(dat.vals)
```
- Override remote access on Windows by setting SKA to a valid path
  so eng archive data will be found. Confirm that remote access is
  disabled and fetch uses local access::
```
    import os
    os.environ['SKA'] = <path_to_ska_root>
    os.environ.pop('ENG_ARCHIVE', None)
    os.environ.pop('SKA_ACCESS_REMOTELY', None)
    from Ska.engarchive import fetch, remote_access
    fetch.add_logging_handler()
    assert remote_access.access_remotely is False
    dat = fetch.Msid('1wrat', '2018:001', '2018:010')
    print(dat.vals)
```
- Override remote access on non-Windows by setting SKA_ACCESS_REMOTELY to 'True'::
```
    import os
    os.environ['SKA_ACCESS_REMOTELY'] = 'True'
    from Ska.engarchive import fetch, remote_access
    fetch.add_logging_handler()
    assert remote_access.access_remotely is True
    dat = fetch.Msid('tephin', '2018:001', '2018:010')
    print(dat.vals)
```
